keps/sig-api-machinery/2019-02-13-custom-resource-definitions.md

```diff
@@ -0,0 +1,212 @@
+because it is an attempt to retrospectively describe the development of a feature in Kubernetes that 
+already exists. We are doing this because custom resource definitions were created and evolved before the
+KEP process was put in place. Because of this, there is no single place that describes the reasons for
+custom resource definitions and how they should be used in Kubernetes.
+
+## Motivation
+
+The primary motivation for custom resources was to provide a simple easy to use extension mechanism
+for Kubernetes clusters to enable customization by Kubernetes users and to facilitate a rich
+ecosystem of plugins for Kubernetes clusters. If Kubernetes is the kernel, CustomResourceDefinitions
+are intended to be loadable kernel modules.
+
+### Goals
+
+   * Provide an easy to use mechanism for the dynamic addition of new API types to Kubernetes.
+   * Enable API extensibility without significant additional load on operators of Kubernetes clusters.
+   * Enable an ecosystem of plugin providers who can produce and distribute extensions that add value to end-user clusters.
+
+### Non-Goals
+   * Replace all existing built-in API objects with Custom Resources.
+   * Enable every possible extension API. Unsupported patterns should use aggregated API servers.
+   * Develop an arbitrary key-value store for applications. Custom resources are control-plane objects.
+   * Develop tooling for managing the installation ordering or dependencies of custom resources. This is deferred to other KEPs.
+   * Expressing or managing inheritance, referential integrity, or other inter-object relationships.
+   * Document how to create a custom resource definition. This is done [elsewhere](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/)
+
+
+## Implementation History
+
+Third-party resources were originally [proposed](https://github.com/kubernetes/kubernetes/pull/11781) 
+shortly after Kubernetes 1.0. At the time it was clear that making the Kubernetes API server
+extensible was critical to the long-term health of the project and there were two different
+approaches for extension proposed:
+   * Plugin resources (initial "ThirdPartyResource" eventually "CustomResourceDefinitions")
+   * Delgation to an "aggregated API server"
+
+ThirdPartyResources were added to the API server, but the API server was actually really not ready
+for them, much of the dynamic discovery and grouping mechanisms were primitive, or non-existent.
+As a result, and because there was little initial support from the core API machinery maintainers
+for ThirdPartyResources, the feature languished with bugs, inconsistencies and missing functionality.
+
+Despite this, ThirdPartyResources gained popularity, especially with the development of the Operator
+pattern by CoreOS. This usage lead to many bug fixes and proving out the model as a viable method for
+developing a rich ecosystem of APIs.
```

Comment:
There was a moment around Feb 2017 when the API Machinery community was trying to decide whether to continue to focus effort on Aggregated API Servers, or to shift effort to TPRs/CRDs.  The community decided on  the latter.  [Here is one of the documents that, at least for me, worked through this decision.](https://docs.google.com/document/d/1y16jKL2hMjQO0trYBJJSczPAWj8vAgNFrdTZeCincmI/edit)